Userland/Libraries/LibGfx/AntiAliasingPainter.cpp

```diff
@@ -169,3 +170,218 @@ void Gfx::AntiAliasingPainter::draw_cubic_bezier_curve(const FloatPoint& control
+    int delta_f2 { subpixel_resoulution * subpixel_resoulution },
+        delta2_f2 { -delta_f2 - delta_f2 };
+
+    /* v/subpixel_resoulution = percentage of pixel intersected by circle */
+    int v { 0 }, v_old { 0 };
+
+    auto predict = [&] {
+        delta_y += delta2_y;
+        y_hat = y + delta_y; /* y_hat is the predicted value of f(i) */
+    };
+
+    auto minimize = [&] {
+        /* Initialize the minimization */
+        delta_f2 += delta2_f2;
+        f2 += delta_f2;
+
+        int e = y_hat * y_hat - f2;
+        int d = 1;
+        y = y_hat;
+
+        /* Force e negative */
+        if (e > 0) {
+            e = -e;
+            d = -1;
+        }
+
+        /* Minimize */
+        int e_old;
+        while (e < 0) {
+            y = y + d;
+            e_old = e;
+            e += y + y - d;
+        }
+
+        /* e or e_old is minimum squared error */
+        if (e + e_old > 0)
+            y -= d;
+    };
+
+    auto correct = [&] {
+        int error = y - y_hat;
+        delta2_y += error;
+        delta_y += error;
+    };
+
+    auto pixel = [&](int x, int y, int alpha) {
+        if (alpha <= 0 || alpha > 255)
+            return;
+        auto pixel_colour = color;
+        pixel_colour.set_alpha(alpha);
+        m_underlying_painter.set_pixel(center + IntPoint { x, y }, pixel_colour, true);
+    };
+
+    auto fill = [&](int x, int ymax, int ymin, int alpha) {
+        while (ymin <= ymax) {
+            pixel(x, ymin, alpha);
+            ymin += 1;
+        }
+    };
+
+    auto eight_pixel = [&](int x, int y, int alpha) {
+        pixel(x, y, alpha);
+        pixel(x, -y - 1, alpha);
+        pixel(-x - 1, -y - 1, alpha);
+        pixel(-x - 1, y, alpha);
+        pixel(y, x, alpha);
+        pixel(y, -x - 1, alpha);
+        pixel(-y - 1, -x - 1, alpha);
+        pixel(-y - 1, x, alpha);
+    };
+
+    while (i < q) {
+        predict();
+        minimize();
+        correct();
+        v_old = v;
+        v += delta_y;
+        if (v >= 0) {
+            /* Single pixel on perimeter */
+            eight_pixel(i, q, (v + v_old) >> 1);
+            fill(i, q - 1, -q, intensity);
+            fill(-i - 1, q - 1, -q, intensity);
+        } else {
+            /* Two pixels on perimeter */
+            v += subpixel_resoulution;
+            eight_pixel(i, q, v_old >> 1);
+            q -= 1;
+            fill(i, q - 1, -q, intensity);
+            fill(-i - 1, q - 1, -q, intensity);
+            if (i < q) {
+                /* Haven't gone below the diagonal */
+                eight_pixel(i, q, (v + subpixel_resoulution) >> 1);
+                fill(q, i - 1, -i, intensity);
+                fill(-q - 1, i - 1, -i, intensity);
+            } else {
+                /* Went below the diagonal, fix v for final pixels */
+                v += subpixel_resoulution;
+            }
+        }
+        i += 1;
+    }
+
+    /* Fill in 4 remaning pixels */
+    int alpha = v >> 1;
```

Comment:
Is the `>> 1` part from the paper? It's much less mental overhead to just have `/ 2`, which any decent compiler would optimize down to the same thing.

Reply:
It was, but I assume that must of been because it's a pretty old paper (1980s) so they could not rely on the compiler as much, silly to keep it though -- I'll fix that